### PR TITLE
ISPN-4955 JCache implementation over HotRod

### PR DESCRIPTION
**Pull request by isavin**
https://issues.jboss.org/browse/ISPN-4955

### CODE REVIEW

**Comment by galderz** [comment]
@isavin Thanks a lot for working on this :). I'll have an in-depth look at the beginning of next week.

File: jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCache.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+   }
+
+   protected static CacheDecorator IDENTITY_CACHE_DECORATOR = new CacheDecorator() {
+      @Override
+      public <K, V> BasicCache<K, V> decorate(BasicCache<K, V> cache) {
+         return cache;
+      }
+   };
+
+   protected static interface CacheDecorator {
```
Comment (galderz) [inline jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCache.java]: Hmmm, I'm not sure I understand what you are trying to do with this decorator. I see that you have used to pass in `FORCE_RETURN_VALUE` flag to the remote cache, but you could have simply configured the RemoteCacheManager to have `infinispan.client.hotrod.force_return_values` set to true on construction (or equivalent in programmatic Hot Rod client configuration). If you do that, I don't think you'd need a decorator.

File: jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java [hunk 2]
```diff
@@ -30,10 +31,10 @@
  * @author Galder Zamarreño
  * @since 5.3
  */
-public class JCacheNotifier<K, V> {
+public abstract class AbstractJCacheNotifier<K, V> {
```
Comment (galderz) [inline jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java]: I'm not sure I understand why you need to create an abstract class for this and also one for `JCacheListenerAdapter`. The two subclasses of AbstractJCacheNotifier are pretty much identical. The real difference comes in `JCacheListenerAdapter`. Also, consider whether you could use delegation instead of inheritance to solve this? IOW, you could just pass the adapter you need as parameter on construction to JCacheNotifier.

File: jcache/commons/src/main/java/org/infinispan/jcache/annotation/AbstractInjectedCacheResolver.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.infinispan.jcache.annotation;
+
+import java.lang.annotation.Annotation;
+
+import javax.cache.Cache;
+import javax.cache.annotation.CacheInvocationContext;
+import javax.cache.annotation.CacheResolver;
+
+public abstract class AbstractInjectedCacheResolver implements CacheResolver {
```
Comment (galderz) [inline jcache/commons/src/main/java/org/infinispan/jcache/annotation/AbstractInjectedCacheResolver.java]: Why create this class?

Comment (galderz) [inline jcache/commons/src/main/java/org/infinispan/jcache/annotation/AbstractInjectedCacheResolver.java]: You can leave it out until you really need it.

File: jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java [hunk 4]
```diff
@@ -371,7 +371,9 @@ public T call() throws Exception {
    @Override
    public <T> Map<K, EntryProcessorResult<T>> invokeAll(
          Set<? extends K> keys, EntryProcessor<K, V, T> entryProcessor, Object... arguments) {
-      checkNotClosed().checkNull(entryProcessor, "entryProcessor").verifyKeys(keys);
```
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java]: Why did you change this? What is wrong with it?

File: jcache/remote/pom.xml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+   <dependencies>
+      <dependency>
+         <groupId>${project.groupId}</groupId>
+         <artifactId>infinispan-jcache-commons</artifactId>
+      </dependency>
+
+      <dependency>
+         <groupId>${project.groupId}</groupId>
+         <artifactId>infinispan-core</artifactId>
```
Comment (galderz) [inline jcache/remote/pom.xml]: You shouldn't need `core` as dependency for the remote JCache implementation. Why did you add it? If you need anything, it should be in `commons`

File: jcache/remote/pom.xml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      <dependency>
+         <groupId>org.kohsuke.metainf-services</groupId>
+         <artifactId>metainf-services</artifactId>
+         <optional>true</optional>
+      </dependency>
+
+      <dependency>
+         <groupId>org.wildfly</groupId>
+         <artifactId>wildfly-controller</artifactId>
+         <optional>true</optional>
```
Comment (galderz) [inline jcache/remote/pom.xml]: Not sure if this dependency can be optional really. Without it you'd not be able to create caches, ...etc, which is needed to pass the TCK.

**Comment by galderz** [comment]
@isavin I've made some comments. It looks very good so far. I've tried to be thorough :)

File: jcache/remote/pom.xml (hunk 5 above)
Comment (isavin) [inline jcache/remote/pom.xml]: I'll remove it. It's not used, must be a left-over.

File: jcache/remote/pom.xml (hunk 6 above)
Comment (isavin) [inline jcache/remote/pom.xml]: This is based on an offline discussion with @tristantarrant 

The TCK requires the ability to create the caches programmatically but usually users would just configure the caches on the server. If they decide to do that they don't need to pull this in as a dependency.

File: jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java (hunk 4 above)
Comment (isavin) [inline jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java]: I moved the null check to the superclass to avoid duplication and I'm planning to do the same for the other checks. I don't think it's required for them to be fluent. They could still be placed on a single line if it's easier to read that way.

File: jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCache.java (hunk 1 above)
Comment (isavin) [inline jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCache.java]: Yes, I'm using it just because withFlags() is not defined in a custom ancestor of the embedded and remote cache. The flags are cleared after the first operation is executed and for a few cases the first operation is not the one requiring the flag.

I've considered the option you're describing but creating a new cache manager and managing it's lifecycle didn't seem right. I'll have another look at this option though as force return seems to be the only flag used.

File: jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java (hunk 2 above)
Comment (isavin) [inline jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java]: The subclasses look identical but createListenerAdapter() returns objects of different classes (same name but different package).

File: jcache/commons/src/main/java/org/infinispan/jcache/annotation/AbstractInjectedCacheResolver.java (hunk 3 above)
Comment (isavin) [inline jcache/commons/src/main/java/org/infinispan/jcache/annotation/AbstractInjectedCacheResolver.java]: It's just a placeholder (no remote implementation yet) but there's a subclass for the embedded cache. Without the abstract class I would need to move org.infinispan.jcache.annotations out of infinispan-jcache-commons.

File: jcache/commons/pom.xml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+   </parent>
+
+   <artifactId>infinispan-jcache-commons</artifactId>
+   <name>Infinispan JCACHE (JSR-107) Common Components</name>
+   <description>Common components for the JCACHE (JSR-107) implementations using Infinispan</description>
+
+   <dependencies>
+      <dependency>
+         <groupId>${project.groupId}</groupId>
+         <artifactId>infinispan-core</artifactId>
```
Comment (isavin) [inline jcache/commons/pom.xml]: Replaced this with a dep on commons.

Split logging.Log to share to avoid duplication and at the same time don't depend on core. One ugly side-effect though is that the _log ids_ are now split between two interfaces (one in commons, one in embedded). Do you have any suggestion on how to make this look less ugly?

**Comment by isavin** [comment]
@galderz thanks for the feedback! I've update the commit based on your feedback. Got rid of the decorator and removed the jcache-commons -> infinispan-core dependency.

File: jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java (hunk 2 above)
Comment (galderz) [inline jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java]: Right, but do you really need to subclass both?

File: jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java (hunk 4 above)
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java]: Keeping them fluent was made on purpose to make it nice and clean, since all operations are validation related. There's nothing wrong with them and you should not modify them for the shake of it really :|

File: jcache/remote/pom.xml (hunk 6 above)
Comment (galderz) [inline jcache/remote/pom.xml]: Ok, makes sense.

**Comment by galderz** [comment]
I've made some further comments. How's the TCK looking so far?

**Comment by galderz** [comment]
Remember to rebase this.

File: jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java (hunk 4 above)
Comment (isavin) [inline jcache/embedded/src/main/java/org/infinispan/jcache/embedded/JCache.java]: OK. I'll revert the change.

**Comment by isavin** [comment]
I have a few more changes locally, the tests still excluded are:
https://github.com/infinispan/infinispan/pull/3264#issuecomment-76967676

Should probably get the loader/writer test failures resolved later today.

**Comment by isavin** [comment]
Updated and rebased. Pending TCK tests: https://github.com/isavin/infinispan/blob/jsr107/jcache/tck-runner/src/test/resources/ExcludeList_remote

File: jcache/remote/src/main/java/org/infinispan/jcache/remote/JCache.java [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+      // Get old value skipping any listeners to impacting
+      // listener invocation expectations set by the TCK.
+      V oldValue = cacheWithCacheStore.get(key);
+
+      MutableJCacheEntry<K, V> mutable = createMutableCacheEntry(oldValue, key);
+      T ret = processEntryProcessor(mutable, entryProcessor, arguments);
+
+      switch (mutable.getOperation()) {
+      case NONE:
+         Thread.dumpStack();
```
Comment (galderz) [inline jcache/remote/src/main/java/org/infinispan/jcache/remote/JCache.java]: You might want to remove these?

File: jcache/remote/src/main/java/org/infinispan/jcache/remote/RemoteCacheWithCacheStore.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+   @SuppressWarnings("unchecked")
+   @Override
+   public V get(Object key) {
+      V value = super.get(key);
+      if (value == null) {
+         try {
+            Thread.dumpStack();
+            value = loadFromCacheLoader((K) key);
+            System.out.println(String.format("foobar k: %s v: %s ", key, value));
```
Comment (galderz) [inline jcache/remote/src/main/java/org/infinispan/jcache/remote/RemoteCacheWithCacheStore.java]: And this too?

**Comment by isavin** [comment]
Ups. Removed the debug msgs.

**Comment by isavin** [comment]
Disabled wildfly-maven-plugin as it breaks on both TeamCity and Jenkins. TCK remote is disabled by default and can be executed manually using -DskipJSR107RemoteTests=false after infinispan-server is started manually.

**Comment by galderz** [comment]
Something's wrong with the pom.xml of TCK:

```
[07:25:32][ERROR]   The project org.infinispan:infinispan-jcache-tck-runner:7.1.0-SNAPSHOT (/mnt/persistent_storage/cloud-user/ispn/buildAgent/work/c9523b7cef7aa565/jcache/tck-runner/pom.xml) has 1 error
[07:25:32][ERROR]     Non-resolvable parent POM: Could not find artifact org.infinispan:infinispan-jcache-parent:pom:7.1.0-SNAPSHOT in redhat-earlyaccess-repository-group (http://maven.repository.redhat.com/earlyaccess/all/) and 'parent.relativePath' points at wrong local POM @ line 7, column 12 -> [Help 2]
```

**Comment by isavin** [comment]
tck-runner was pointing to version 7.1.0-SNAPSHOT of the parent instead of 7.2.0-SNAPSHOT. Fixed.

**Comment by galderz** [comment]
Looks good now, integrating...

**Comment by galderz** [comment]
Integrated

**Comment by galderz** [comment]
JIRA to remain open since more work still to come
